File: 124/D.cpp
//encoding:utf-8
//#include <iostream> //c++
//#include<stack> //スタックに関して　最新のデータから捨てる
//#include<queue> //キューに関して　最古のデータから捨てる
//#include<math.h>
//#include<vector>
//#include<algorithm> //sort とかにいる
#include<bits/stdc++.h>//vector,algorithm,que
#include<time.h> //時間にかんして

#define ll long long
#define MOD 1000000007
using namespace std;
int main(){
  int N,K;
  cin>>N>>K;
  string S;
  cin>>S;

  vector<int> Nums;
  int now = 1;//今みている数
  //もし最初の数字が0ならその分がカウントされる
  int cnt = 0;
  for(int i=0;i<N;i++){
    if(S[i]==(char)('0'+now))cnt++;
    else{
      Nums.push_back(cnt);
      now = 1 - now;
      cnt = 1;
    }
  }

  if(cnt!=0) Nums.push_back(cnt);

  if(Nums.size()%2==0)Nums.push_back(0);

  int Add = 2 * K + 1;
  int ans = 0;
  //1-0-...の1から始めるの
  for(int i=0;i<Nums.size();i+=2){
    int tmp = 0;
    int left = i;
    int right = min(i+Add,(int)Nums.size());
    for(int j=left;j<right;j++){
      tmp += Nums[j];
    }
    ans = max(ans,tmp);
  }

  cout<<ans<<endl;
  return 0;
}

File: 124/D.py
# encoding:utf-8
import copy
import numpy as np
import random
import bisect #bisect_left　これで二部探索の大小検索が行える
import math


N,K = map(int,input().split())
S = str(input())

now,cnt = 1,0
Nums = []
for i in range(N):
    if S[i] == str(now):
        cnt += 1
    else:
        Nums.append(cnt)
        cnt = 1
        now = 1 - now

if cnt!=0:
    Nums.append(cnt)

if len(Nums)%2==0:
    Nums.append(0)

#累積和
"""
Sums = [0]
tmp = 0
for i in range(len(Nums)):
    tmp += Nums[i]
    Sums.append(tmp)

ans = 0
Add = 2*K+1
for i in range(len(Sums)):
    if i%2==0:
        left = i
        right = min(i+Add,len(Sums)-1)
        tmp = Sums[right]-Sums[left]
        ans = max(tmp,ans)
"""
# 尺取り法
ans = 0
add = 2*K+1
left = 0
right = min(0+add,len(Nums))
pre = sum(Nums[left:right])
ans = pre
for i in range(len(Nums)-add):
    if i%2==0:
        left = i
        right = min(i+add,len(Nums)-1)
        pre += sum(Nums[right:right+2])-sum(Nums[left:left+2])
        ans = max(pre,ans)

print(ans)
